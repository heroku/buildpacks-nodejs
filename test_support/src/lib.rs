use libcnb_test::{
    assert_contains, BuildConfig, BuildpackReference, ContainerConfig, TestContext, TestRunner,
};
use std::fmt;
use std::fmt::Formatter;
use std::time::Duration;

const PORT: u16 = 8080;
const TIMEOUT: u64 = 10;

pub enum Builder {
    Heroku20,
    Heroku22,
}

impl fmt::Display for Builder {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        match *self {
            Builder::Heroku20 => write!(f, "heroku/buildpacks:20"),
            Builder::Heroku22 => write!(f, "heroku/builder:22"),
        }
    }
}

pub fn get_function_invoker_buildpacks() -> Vec<BuildpackReference> {
    vec![
        BuildpackReference::Other(String::from("heroku/nodejs-engine")),
        BuildpackReference::Other(String::from("heroku/nodejs-npm")),
        BuildpackReference::Crate,
    ]
}

pub fn get_yarn_buildpacks() -> Vec<BuildpackReference> {
    vec![
        BuildpackReference::Other(String::from("heroku/nodejs-engine")),
        BuildpackReference::Crate,
    ]
}

pub fn get_function_invoker_build_config(fixture: &str, builder: Builder) -> BuildConfig {
    BuildConfig::new(
        builder.to_string(),
        format!("../../test/fixtures/{fixture}"),
    )
    .buildpacks(get_function_invoker_buildpacks())
    .to_owned()
}

pub fn get_yarn_build_config(fixture: &str, builder: Builder) -> BuildConfig {
    BuildConfig::new(
        builder.to_string(),
        format!("../../test/fixtures/{fixture}"),
    )
    .buildpacks(get_yarn_buildpacks())
    .to_owned()
}

pub fn test_node_function(fixture: &str, builder: Builder, test_body: fn(TestContext)) {
    TestRunner::default().build(get_function_invoker_build_config(fixture, builder), |ctx| {
        test_body(ctx)
    });
}

pub fn test_yarn_app(fixture: &str, builder: Builder, test_body: fn(TestContext)) {
    TestRunner::default().build(get_yarn_build_config(fixture, builder), |ctx| {
        test_body(ctx)
    });
}

pub fn assert_health_check_responds(ctx: &TestContext) {
    ctx.start_container(ContainerConfig::new().expose_port(PORT), |container| {
        std::thread::sleep(Duration::from_secs(TIMEOUT));

        let addr = container
            .address_for_port(PORT)
            .expect("couldn't get container address");

        let resp = ureq::post(&format!("http://{0}", addr))
            .set("x-health-check", "true")
            .call()
            .expect("request to container failed")
            .into_string()
            .expect("response read error");

        assert_contains!(resp, "OK")
    })
}

pub fn assert_web_response(ctx: &TestContext, text: &'static str) {
    ctx.start_container(ContainerConfig::new().expose_port(PORT), |container| {
        std::thread::sleep(Duration::from_secs(5));

        let addr = container
            .address_for_port(PORT)
            .expect("couldn't get container address");

        let resp = ureq::get(&format!("http://{}/", addr))
            .call()
            .expect("request to container failed")
            .into_string()
            .expect("response read error");

        assert_contains!(resp, text);
    })
}
