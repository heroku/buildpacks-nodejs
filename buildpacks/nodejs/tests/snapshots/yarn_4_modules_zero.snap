---
source: test_support/src/lib.rs
---
===> ANALYZING
Image with name "<image-name>" not found
===> DETECTING
heroku/nodejs-engine   <buildpack-version>
heroku/nodejs-corepack <buildpack-version>
heroku/nodejs-yarn     <buildpack-version>
===> RESTORING
<restoring-output>
===> BUILDING

## Heroku Node.js Engine

- Checking Node.js version
  - Detected Node.js version range: `>=18.0.0 <19.0.0-0`
  - Resolved Node.js version: `18.20.8`
- Installing Node.js distribution
  - GET https://nodejs.org/download/release/v18.20.8/node-v18.20.8-<arch>.tar.gz ... (OK)
  - Downloading ... (<time_elapsed>)
  - Verifying checksum
  - Extracting Node.js `18.20.8 (<arch>)`
  - Installing Node.js `18.20.8 (<arch>)` ... (<time_elapsed>)
- Done (finished in <time_elapsed>)

## Heroku Node.js Corepack

- Using Corepack version `0.32.0`
- Found `packageManager` set to `yarn@4.0.0` in `package.json`
- Enabling Corepack
  - Creating Corepack shims
  - Executing `corepack enable --install-directory /layers/heroku_nodejs-corepack/shim/bin yarn`
- Installing `yarn@4.0.0`
  - Creating Corepack package manager
  - Running `corepack prepare`

      Preparing yarn@4.0.0...

  - Done (<time_elapsed>)
- Done (finished in <time_elapsed>)

## Heroku Node.js Yarn

- Yarn CLI operating in yarn 4.0.0 mode.
- Setting up yarn dependency cache
  - Running `yarn config set enableGlobalCache false`

      ➤ YN0000: Successfully set enableGlobalCache to false

  - Done (<time_elapsed>)
  - Yarn zero-install detected. Skipping dependency cache.
- Installing dependencies
  - Running `yarn install --immutable --inline-builds --immutable-cache`

      ➤ YN0000: · Yarn 4.0.0
      ➤ YN0000: ┌ Resolution step
      ➤ YN0000: └ Completed
      ➤ YN0000: ┌ Fetch step
      ➤ YN0000: └ Completed
      ➤ YN0000: ┌ Link step
      ➤ YN0000: └ Completed
      ➤ YN0000: · Done in <time_elapsed>

  - Done (<time_elapsed>)
- Running scripts
  - No build scripts found
- Pruning dev dependencies
  - Running `yarn heroku prune`

      ➤ YN0000: ┌ Resolution step
      ➤ YN0085: │ - dotenv@npm:16.6.1
      ➤ YN0000: └ Completed
      ➤ YN0000: ┌ Fetch step
      ➤ YN0000: └ Completed
      ➤ YN0000: ┌ Link step
      ➤ YN0000: └ Completed
      ➤ YN0000: Done in <time_elapsed>

  - Done (<time_elapsed>)
- Done (finished in <time_elapsed>)
===> EXPORTING
Adding layer 'heroku/nodejs-engine:available_parallelism'
Adding layer 'heroku/nodejs-engine:dist'
Adding layer 'heroku/nodejs-engine:web_env'
Adding layer 'heroku/nodejs-corepack:mgr'
Adding layer 'heroku/nodejs-corepack:shim'
Adding layer 'buildpacksio/lifecycle:launch.sbom'
Added 1/1 app layer(s)
Adding layer 'buildpacksio/lifecycle:launcher'
Adding layer 'buildpacksio/lifecycle:config'
Adding layer 'buildpacksio/lifecycle:process-types'
Adding label 'io.buildpacks.lifecycle.metadata'
Adding label 'io.buildpacks.build.metadata'
Adding label 'io.buildpacks.project.metadata'
Setting default process type 'web'
Saving <image-name>...
*** Images (<random-hex>):
      <image-name>
Adding cache layer 'heroku/nodejs-engine:dist'
Adding cache layer 'heroku/nodejs-corepack:mgr'
Adding cache layer 'heroku/nodejs-corepack:shim'
Successfully built image '<image-name>'
